src/backend/dx12/src/root_constants.rs

```diff
@@ -0,0 +1,302 @@
+//! Processing push constant ranges
+//!
+//! This module provides utitlity functions to make push constants root signature
+//! compatible. Root constants are non-overlapping, therefore, the push constant
+//! ranges passed at pipeline layout creation need to be `split` into disjunct
+//! ranges. The disjunct ranges can be then converted into root signature entries.
+
+use hal::pso;
+use std::cmp::Ordering;
+use std::ops::Range;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct RootConstant {
+    pub stages: pso::ShaderStageFlags,
+    pub range: Range<u32>,
+}
+
+impl RootConstant {
+    fn is_empty(&self) -> bool {
+        self.range.end <= self.range.start
+    }
+
+    // Divide a root constant into two separate ranges depending on the overlap
+    // with another root constant.
+    fn divide(self, other: &RootConstant) -> (RootConstant, RootConstant) {
+        let left = RootConstant {
+            stages: self.stages,
+            range: self.range.start..other.range.start,
+        };
+
+        let right = RootConstant {
+            stages: self.stages,
+            range: other.range.start..self.range.end,
+        };
+
+        (left, right)
+    }
+}
+
+impl PartialOrd for RootConstant {
+    fn partial_cmp(&self, other: &RootConstant) -> Option<Ordering> {
+        Some(self.range.start.cmp(&other.range.start)
+            .then(self.range.end.cmp(&other.range.end))
+            .then(self.stages.cmp(&other.stages)))
+    }
+}
+
+impl Ord for RootConstant {
+    fn cmp(&self, other: &RootConstant) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+impl<'a> From<&'a (pso::ShaderStageFlags, Range<u32>)> for RootConstant {
+    fn from(&(stages, ref range): &'a (pso::ShaderStageFlags, Range<u32>)) -> Self {
```

Comment:
where is this one needed?

Reply:
It's used in `into_vec`.

Reply:
Let's move this code into the `into_vec` implementation then. I don't think this `From<>` is a good thing to have here.